enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+  - "@benluddy"
+  - "@ecordell"
+  - "@kevinrizza"
+  - "@njhale"
+approvers:
+  - "@kevinrizza"
+  - "@dmesser"
+  - "@ecordell"
+  - "@shawn-hurley"  
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: provisional
+see-also:
+  - "N/A"
+replaces:
+  - "N/A"
+superseded-by:
+  - "N/A"
+---
+
+# olm-deletes-operands
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The purpose of this enhancement is to outline how the [Operator Lifecycle Manager (OLM)][olm] can let operator users opt-in to the clean up of their operands, or operator workloads, upon the uninstallation of their operator. Specifically when uninstalling an operator, OLM would first delete all CustomResources (CRs) managed by an operator to provide the operator time to clean up all associated resources before deleting the operator itself.
+
+## Motivation
+
+By design when OLM uninstalls an operator it does not remove any of the operator’s owned CustomResourceDefinitions (CRDs) and CRs in order to prevent data loss.
+The operator user or cluster admin is expected to manually delete CRDs, CRs and any other secondary resources created by the operator.
+While some secondary resources that have their [OwnerReference][owner-reference] set to a CR would be garbage collected, others such as off-cluster or cross-namespaced resources with no owner references would remain orphaned.
+
+Without the operator around to process deletion events of CRs, it becomes hard to cleanly remove deployed resources, especially when they exist outside of the cluster.
+
+### Goals
+
+- Clean up includes removal of CRs (operands) and CRDs
+- Operands and CRDs are not automatically removed
+- Upon operator uninstall a user can opt-in to the clean up behavior
+- The operand removal should have a timeout
+- After the operand removal timeout the operator should still be uninstalled
+- Exceeding the operand removal timeout should create an alert event
+
+### Non-Goals
+
+- OLM handles cleanup of anything other than CRs and CRDs
+- Handle custom cleanup logic like filtering/ordering the deletion of CRs 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator user I can instruct OLM to clean up all of my operator's CRs and CRDs when I uninstall my operator.
+
+#### Story 2
+
+As an operator user I want to see an alert event when OLM does not succeed in operand cleanup after a configured timeout.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Specify cleanup via Subscription
+
+To provide a way to opt-in for operand cleanup, the [`Subscription`][subscription] spec can include an optional field `spec.operandCleanup` which lets a user configure the cleanup behaviour.
+
+OLM would initiate cleanup upon the deletion of a `Subscription` instance, based on `spec.operandCleanup`:
```

Comment:
What about CSVs that aren't installed by `Subscriptions`?

Reply:
Hmm, that's tricky. If the cleanup option was placed on the `CSV` object then that would cover the case for `Subscription`-less installs.

And in that case I'm thinking it might make more sense to initiate clean up when the user deletes the `CSV` object since that's actually the final step to uninstalling an operator.

But that would degrade the UX for the presumably common case of installing via `Subscription`, where you have to update the `CSV` after it's created:
```sh
$ kubectl create -f subscription.yaml
$ kubectl patch csv <csv-name> -p '{"spec":{"cleanup":{true}}'
```

So we could just mirror the `spec.cleanup` field from the `Subscription` to the `CSV`, where `subscription.spec.cleanup` overwrites `csv.spec.cleanup`.
That way the UX for opting to clean up when installing via `Subscription` is still simple, and it also covers the case of CSVs not installed with `Subscriptions`.
Plus it also covers the use case of allowing a default cleanup policy set via the CSV (https://github.com/operator-framework/enhancements/pull/46#discussion_r483673349).

WDYT @njhale @ecordell ?

Reply:
SGTM

Reply:
Good point, though I am a little worried this would create the scenario where somehow the cleanup toggles get out of sync between `Subscription` and `CSV`. I'd lean towards only `Subscription` controlling the cleanup behavior as it also controls the update behavior of the CSV.

Reply:
@dmesser As I've been writing this out, it seems to me that the CSV should be the only place to have the operand cleanup config.

Having the clean up knob only be on the Subscription is a problem because it doesn't cover a subscription-less install case.

Also since deleting the CSV is the actual step in uninstall which effectively removes the operator, the UX that makes more sense to me is "opt-in before uninstall" rather than "opt-in at install time". The former is more flexible in allowing opt-in later even if it wasn't configured at install time.

So the CSV rather than the subscription might be the more logical point of having this option. 
Plus it also covers the use case of letting an operator author set the default cleanup option in the CSV which can be overridden later by the user before uninstall time. 

I've got an incoming update that reframes the proposal to make the CSV only case clearer.
This and other points: like no timeouts by default, adding a finalizer to the CSV to hold off operator deletion during operand cleanup, and showing operand cleanup status on the CSV (useful when diagnosing a stalled cleanup process).
Hopefully that should give us a better ground for deciding Subscription vs CSV (or both).

Reply:
> Plus it also covers the use case of letting an operator author set the default cleanup option in the CSV which can be overridden later by the user before uninstall time.

Are we sure we want to give operator authors the ability to make operand cleanup an opt-out for OLM admins? If we do, then OLM admins may be surprised and upset that deleting a CSV all of a sudden causes operands to be deleted when their past experience led them to believe CSV deletion is a safe operation.